engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java

```diff
@@ -2302,8 +2302,11 @@ protected void migrate(final VMInstanceVO vm, final long srcHostId, final Deploy
             StoragePoolVO targetPool = _storagePoolDao.findById(poolId);
             StoragePoolVO currentPool = _storagePoolDao.findById(volume.getPoolId());
 
-            if (_poolHostDao.findByPoolHost(targetPool.getId(), host.getId()) == null) {
-                throw new CloudRuntimeException(String.format("Cannot migrate the volume [%s] to the storage pool [%s] while migrating VM [%s] to target host [%s]. The host does not have access to the storage pool entered.", volume.getUuid(), targetPool.getUuid(), profile.getUuid(), host.getUuid()));
+            executeManagedStorageChecks(targetHost, currentPool, volume);
+            if (_poolHostDao.findByPoolHost(targetPool.getId(), targetHost.getId()) == null) {
+                throw new CloudRuntimeException(
+                        String.format("Cannot migrate the volume [%s] to the storage pool [%s] while migrating VM [%s] to target host [%s]. The host does not have access to the storage pool entered.",
+                                volume.getUuid(), targetPool.getUuid(), profile.getUuid(), targetHost.getUuid()));
             }
             if (currentPool.getId() == targetPool.getId()) {
```

Comment:
You mean, I should not add an entry to the map if the volume is already in the "targetPool"?

I did this way because of XenServer; when we have VMs with more than one volume. Then, if one volume is in a shared storage and the other in local storage, we need to send to Xen the complete map of “volume-storagePool”. Otherwise, the migration will not work. Since I am mapping the volume to its current storage pool I am expecting the other hypervisors to simply ignore the entry.

Reply:
Yeah, in the old code, the volume was not added to the map (even for XenServer) unless it needed to be migrated to a new storage pool. It sounds like you are saying that logic was incorrect, though?

Reply:
Yes. It was not working with XenServer before because it requires all disks to be mapped. Even the ones that will not be migrated (e.g. disks in shared storage).